shell/platform/android/io/flutter/plugin/editing/InputConnectionAdaptor.java

```diff
@@ -132,19 +140,59 @@ public boolean setComposingText(CharSequence text, int newCursorPosition) {
   public boolean finishComposingText() {
     boolean result = super.finishComposingText();
 
-    if (Build.VERSION.SDK_INT >= 21) {
-      // Update the keyboard with a reset/empty composing region. Critical on
-      // Samsung keyboards to prevent punctuation duplication.
-      CursorAnchorInfo.Builder builder = new CursorAnchorInfo.Builder();
-      builder.setComposingText(-1, "");
-      CursorAnchorInfo anchorInfo = builder.build();
-      mImm.updateCursorAnchorInfo(mFlutterView, anchorInfo);
+    // Apply Samsung hacks. Samsung caches composing region data strangely, causing text
+    // duplication.
+    if (mIsSamsung) {
+      if (Build.VERSION.SDK_INT >= 21) {
+        // Samsung keyboards don't clear the composing region on finishComposingText.
+        // Update the keyboard with a reset/empty composing region. Critical on
+        // Samsung keyboards to prevent punctuation duplication.
+        CursorAnchorInfo.Builder builder = new CursorAnchorInfo.Builder();
+        builder.setComposingText(-1, "");
+        CursorAnchorInfo anchorInfo = builder.build();
+        mImm.updateCursorAnchorInfo(mFlutterView, anchorInfo);
+      }
+      // TODO(garyq): There is still a duplication case that comes from hiding+showing the keyboard.
```

Comment:
Hiding and showing the keyboard can restart the IMM. May be related?

Reply:
Don't have a solid answer here, but in my experimentation, restarting IMM did not seem to have any noticeable effect on this behavior.